lib/AST/ASTScopeCreation.cpp

```diff
@@ -0,0 +1,1061 @@
+//===--- ASTScopeCreation.cpp - Swift Object-Oriented AST Scope -----------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// This file implements the creation methods of the ASTScopeImpl ontology.
+///
+//===----------------------------------------------------------------------===//
+#include "swift/AST/ASTScope.h"
+
+#include "swift/AST/ASTContext.h"
+#include "swift/AST/ASTWalker.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/Expr.h"
+#include "swift/AST/Initializer.h"
+#include "swift/AST/LazyResolver.h"
+#include "swift/AST/Module.h"
+#include "swift/AST/ParameterList.h"
+#include "swift/AST/Pattern.h"
+#include "swift/AST/Stmt.h"
+#include "swift/AST/TypeRepr.h"
+#include "swift/Basic/STLExtras.h"
+#include "llvm/Support/Compiler.h"
+#include <algorithm>
+
+using namespace swift;
+namespace swift {
+namespace ast_scope {
+
+#pragma mark DeferredNodes
+
+/// \c SourceFiles and \c BraceStmts have \c Decls and \c ASTNodes in them that
+/// must be scoped in distant descendants.
+/// As scopes are created, pass down the \c ASTNodes in enclosing \c Decls
+/// that really need to be in subscopes.
+///
+/// Because parsing and lookup (typechecking) are interleaved for some files,
+/// we need to be able to resume scope tree creation. \c DeferredNodes also
+/// helps with this by remembering the \c lastAdopter.
+
+class DeferredNodes final {
+  /// The nodes that must be passed down to deeper levels in the scope tree.
+  std::deque<ASTNode> nodes;
```

Comment:
Do you need to use an std::deque? I have heard of it having sub-par performance IIRC.

Reply:
Also, another thing couldn't the DeferredNodes be committed as a separate commit? It is separable, no?

Reply:
What would be the benefit in committing it separately? It is only used by the new scope code.

Reply:
There was some reason for `deque`, but I'll look at it again.

Reply:
Good, thanks! A vector will do.

Reply:
The benefit is by splitting up the patch into smaller pieces, it becomes easier to review. A big blob like this has historically resulted in us missing some pretty bad bugs.

Reply:
Keep in mind, I am /not/ talking about committing it separately, I am talking about as a separate commit in the same PR. So you can have multiple commits in the PR that make it easy to review the code separately.

Reply:
Oh, no I did not understand that. I misread `splitting up the patch into smaller pieces` to mean `splitting up the patch into smaller pull requests`. I can split it up into multiple commits, but if each one is non-functional till the final one, how does that help? If you are reviewing a comment that introduces DeferredNodes, but that type isn't used anywhere, how would it be reviewed?

On the other hand, I do think it's worth while to split off the trailing-angle-bracket-location-for-EditorPlaceholders as a separate PR, so that folks like @rintaro can look at that in isolation.
It homes in on his expertise and the added functionality is self-evident.

Reply:
I agree that when each commit makes functional sense, it's easier to spot bugs. In this case, I'm not so sure because if DeferredNodes does the wrong thing, there's no way to know without reading the very code I would be splitting it off from.

However, I do think it might well serve, since DeferredNodes is declared in a .cpp file, to move the method bodies inside of the class definition.

@slavapestov ? What do you think? You're going to be reviewing for functionality, so it's important to accomadate you. @DougGregor, too. (@DougGregor `DeferredNodes` is my stab at more readable what-you-called "continuations", without the laziness constraint.